### PR TITLE
 Support for escaped chars in U+1xxxx range. 

### PR DESCRIPTION
**Pull request by xxxbxxx**
using the utf-16 surrogate pairs

### CODE REVIEW

File: json.c [hunk 1]
```diff
@@ -343,25 +344,42 @@ static int json_get_string_size(struct json_parse_state_s *state,
         //      4       21      U + 10000       U + 10FFFF      11110xxx        10xxxxxx        10xxxxxx        10xxxxxx
         // Note: the high and low surrogate halves used by UTF-16 (U+D800 through U+DFFF) and code points not encodable by UTF-16 (those after U+10FFFF) are not legal Unicode values, and their UTF-8 encoding must be treated as an invalid byte sequence.
     
-        if (codepoint <= 0x7f) {
-          offset += 1;
-          data_size += 1;
+        if (high_surrogate_codepoint) {
```
Comment (sheredom) [inline json.c]: I don't really know what a `high_surrogate_codepoint` is - could you add some comments?

File: json.c [hunk 2]
```diff
@@ -343,25 +344,42 @@ static int json_get_string_size(struct json_parse_state_s *state,
...
+        else if (codepoint <= 0x7f) {
+          data_size += 0;
         }
         else if (codepoint <= 0x7ff) {
-          offset += 2;
-          data_size += 2;
+          data_size += 1;
+        }
+        else if (codepoint >= 0xd800 && codepoint <= 0xdbff) {
+          if (offset+7 > size || '\\' != src[offset+5] || 'u' != src[offset+6]) {
```
Comment (sheredom) [inline json.c]: It doesn't look like you've ran clang-format on this - the styling is off!

File: json.c [hunk 3]
```diff
@@ -804,6 +822,7 @@ static int json_get_number_size(struct json_parse_state_s *state) {
       if (json_parse_flags_allow_equals_in_object & flags_bitset) {
         break;
       }
+      // FALLTHROUGH
```
Comment (sheredom) [inline json.c]: Gurgh - is this to stop a compiler warning?

**Comment by xxxbxxx** [comment]
hi!

No, indeed, this is not about supporting utf-16, but dealing with this case:

> To escape an extended character that is not in the Basic Multilingual Plane, the character is represented as a twelve-character sequence, encoding the UTF-16 surrogate pair. So, for example, a string containing only the G clef character (U+1D11E) may be represented as "\uD834\uDD1E".

https://tools.ietf.org/id/draft-ietf-json-rfc4627bis-09.html#rfc.section.7

(I not sure there are really any good reasons to want to escape such characters, but some other libraries (like the one in php IIRC), that do this. So the idea was to be able to accept it.)

File: json.c (hunk 3 above)
Comment (xxxbxxx) [reply json.c]: yep. with gcc 8.2.1, when doing `cd test; cmake .; make`
the build fails with 

```json.c: In function 'json_get_number_size':
json.c:799:10: error: this statement may fall through [-Werror=implicit-fallthrough=]
       if (json_parse_flags_allow_equals_in_object & flags_bitset) {
          ^
json.c:802:5: note: here
     default:
     ^~~~~~~
cc1: all warnings being treated as errors
```

**Comment by sheredom** [comment]
Ok cool! If you could fix up the two remaining comments then I'll happily merge this! Thanks for contributing 😄

**Pull request merged**